sign/eddilithium3/eddilithium.go

```diff
@@ -0,0 +1,210 @@
+// eddilithium3 implements the hybrid signature scheme Ed25519-Dilithium3.
+package eddilithium3
+
+import (
+	"crypto"
+	cryptoRand "crypto/rand"
+	"errors"
+	"io"
+
+	"github.com/cloudflare/circl/internal/shake"
+	"github.com/cloudflare/circl/sign/dilithium/mode3"
+	"github.com/cloudflare/circl/sign/ed25519"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	SeedSize = mode3.SeedSize // = ed25519.SeedSize = 32
+
+	// Size of a packed PublicKey
+	PublicKeySize = mode3.PublicKeySize + ed25519.PublicKeySize
+
+	// Size of a packed PrivateKey
+	PrivateKeySize = mode3.PrivateKeySize + ed25519.PrivateKeySize
+
+	// Size of a signature
+	SignatureSize = mode3.SignatureSize + ed25519.SignatureSize
+)
+
+// PublicKey is the type of an EdDilithium3 public key.
+type PublicKey struct {
+	e ed25519.PublicKey
+	d mode3.PublicKey
+}
+
+// PrivateKey is the type of an EdDilithium3 private key.
+type PrivateKey struct {
+	e ed25519.KeyPair
+	d mode3.PrivateKey
+}
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GenerateKey(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	var seed [SeedSize]byte
+	if rand == nil {
+		rand = cryptoRand.Reader
+	}
+	_, err := io.ReadFull(rand, seed[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pk, sk := NewKeyFromSeed(&seed)
+	return pk, sk, nil
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[SeedSize]byte) (*PublicKey, *PrivateKey) {
```

Comment:
In Ed22519, this seed is actually the private key. What are the implications of using the same seed for both?

Reply:
We're not using the same seed for both: we're deriving two different seeds from a single seed.  Both Dilithium and Ed25519 don't use their respective seeds directly, but stretch them in a similar way to derive the full private keys.  Even if there is a catastrophic weakness in Dilithium which would allow one to obtain the real private key, then one still isn't able to derive the seed as that would require breaking Shake (which Dilithium uses internally.). Thus it would actually be safe to use literally the same seed (instead of deriving two subseeds,) but that would set a bad example if someone every copy and pasted this for other schemes which don't work like that. So, this is safe and actually a bit conservative.

Reply:
I see, this argument should be documented somewhere.

Reply:
Done.